libr/bin/format/pe/pe.c

```diff
@@ -2736,37 +2800,93 @@ R_API void PE_(bin_pe_parse_resource)(struct PE_(r_bin_pe_obj_t) *bin) {
 	_store_resource_sdb (bin);
 }
 
-static void bin_pe_get_certificate(struct PE_ (r_bin_pe_obj_t) * bin) {
-	ut64 size, vaddr;
-	ut8 *data = NULL;
-	int len;
+static int bin_pe_init_security(struct PE_(r_bin_pe_obj_t) * bin) {
+	Pe_image_security_directory *security_directory;
+	PE_(image_data_directory) *data_dir_security;
+	PE_DWord paddr, offset;
+	Pe_certificate* cert;
+	ut32 size;
 	if (!bin || !bin->nt_headers) {
-		return;
+		return false;
 	}
-	bin->cms = NULL;
-	size = bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_SECURITY].Size;
-	vaddr = bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_SECURITY].VirtualAddress;
-	if (size < 8) {
-		return;
+	if (bin->nt_headers->optional_header.NumberOfRvaAndSizes < 5) {
+		return false;
 	}
-	data = calloc (1, size);
-	if (!data) {
-		return;
+	data_dir_security = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_SECURITY];
+	paddr = bin_pe_rva_to_paddr (bin, data_dir_security->VirtualAddress);
+	size = data_dir_security->Size;
+	if (size < 8 || paddr > bin->size || paddr + size > bin->size) {
+		bprintf ("Invalid certificate table");
+		return false;
 	}
-	if (vaddr > bin->size || vaddr + size > bin->size) {
-		bprintf ("vaddr greater than the file\n");
-		free (data);
-		return;
+
+	security_directory = R_NEW0 (Pe_image_security_directory);
+	if (!security_directory) {
+		r_sys_perror ("R_NEW0 (security_directory)");
+		return false;
 	}
-	//skipping useless header..
-	len = r_buf_read_at (bin->b, vaddr + 8, data, size - 8);
-	if (len < 1) {
-		R_FREE (data);
-		return;
+	bin->security_directory = security_directory;
+
+	offset = paddr;
+	while (offset < paddr + size) {
+		Pe_certificate** tmp = (Pe_certificate **) realloc (security_directory->certificates, (security_directory->length + 1) * sizeof(cert));
+		if (!tmp) {
+			r_sys_perror ("realloc (certificates)");
+			return false;
+		}
+		cert = R_NEW0 (Pe_certificate);
+		if (!cert) {
+			r_sys_perror ("R_NEW0 (security_directory)");
```

Comment:
no need to eprintf the error in every alloc function unless its something that can fail on normal situations